src/YPropertyEditor.cc

```diff
@@ -0,0 +1,281 @@
+
+#include <YPropertyEditor.h>
+
+#include <YUI.h>
+#include <YWidgetFactory.h>
+#include <YDialog.h>
+#include <YLayoutBox.h>
+#include <YAlignment.h>
+#include <YButtonBox.h>
+#include <YPushButton.h>
+#include <YEvent.h>
+#include <YPopupInternal.h>
+#include <YComboBox.h>
+#include <YInputField.h>
+#include <YIntField.h>
+
+#include <limits>
+
+#define YUILogComponent "ui-property-editor"
+#include "YUILog.h"
+
+class YPropertyEditorPriv
+{
+public:
+
+    YPropertyEditorPriv(YWidget * widget) : _widget(widget),
+        popup(nullptr), combo(nullptr), intfield(nullptr),
+        input(nullptr), okButton(nullptr), cancelButton(nullptr)
+    {}
+
+    bool edit(const std::string &property);
+
+private:
+
+    /**
+     * show the dialog on the screen
+     * @param property Name of the property to edit
+     */
+    void show(const std::string &property);
+
+    /**
+     * Run the main event loop
+     * @param  property Name of the property to edit
+     * @return true if the value has been changed, false otherwise
+     */
+    bool run(const std::string &property);
+
+    /**
+     * Close the dialog window
+     */
+    void close();
+
+    /**
+     * Refresh the dialog conatining the widget
+     */
+    void refreshDialog();
+
+    YWidget * _widget;
+
+    /**
+     * Is the property read-only?
+     * @param  property Name of the property
+     * @return true if it is read-only, false if it can be changed
+     */
+    bool isReadOnly(const std::string &property);
+
+    // UI widgets
+    // the main popup
+    YDialog *popup;
+
+    // input widgets
+    YComboBox *combo;
+    YIntField *intfield;
+    YInputField *input;
+
+    // buttons
+    YPushButton *okButton;
+    YPushButton *cancelButton;
+
+    /**
+     * Is the property editable? Editable property is not read-only and
+     * it is String, Integer or Boolean type.
+     * @param  property Name of the property
+     * @return true if the property can be changed
+     */
+    bool editable(const std::string &property);
+};
+
+YPropertyEditor::YPropertyEditor(YWidget * widget)
+: priv(new YPropertyEditorPriv(widget))
+{
+}
+
+YPropertyEditor::~YPropertyEditor()
+{
+}
```

Comment:
We're interleaving the method definitions for YPropertyEditorPriv and YPropertyEditor, that is confusing.